### DISCUSSION TITLE
Getting a transient object's primary key (when possible)

### DISCUSSION THREAD

**Discussion by sh-at-cs** [sqlalchemy/sqlalchemy#8566] [Usage Questions] [answered]
Is there a utility function or anything else in SQLAlchemy that lets me get a **transient** object's primary key?

I realize this isn't always possible, e.g. when the primary key is only defined on the underlying table but not part of the mapped attributes so that SQLAlchemy must let the DB come up with a primary key, but otherwise it should be, shouldn't it?

This could be useful as it would allow one to do things like `session.get(transient_obj.__class__, get_primary_key(transient_obj))` generically without knowing the exact class of `transient_obj` in advance, only that it's a mapped object.

I could probably stitch something together out of [`Mapper.primary_key`](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.Mapper.primary_key) and [`Mapper.get_property_by_column()`](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.Mapper.get_property_by_column), but I thought I'd ask before I waste time on something that already exists.

**Comment by zzzeek** [comment, marked as answer]
yes there's a lesser used function [identity_key](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=identity_key#sqlalchemy.orm.util.identity_key) that does this, it's not as modernistically styled as I'd prefer but easy enough the identity key comes out as:

```py
from sqlalchemy.orm import identity_key

# tuple of (cls, primary_key, identity_token)
# the identity_token is something that you aren't using :) and will be None unless you use the horizontal shard API
#
my_key = identity_key(instance=transient_obj)
mapped_class, pk, identity_token = my_key

# pass all three (identity_token is not important if not using horizontal shard API)
obj = session.get(mapped_class, pk, identity_token=identity_token)

```

note that if you have a transient object with a particular primary key, and that PK is actually in the DB, you can also just merge the object:

```py
persistent_obj = session.merge(my_transient_obj)
```

**Reply by sh-at-cs** [reply]
Thanks a lot!!
